### PR TITLE
Draft of the guide

### PR DESCRIPTION
**Pull request by pnowosie**

### CODE REVIEW

**Review by 83tb — COMMENTED** [review]
@pnowosie 

1) I am missing short birds-eye view todo list for someone that would have to implement this driver, maybe at the end of Overview section. It could go like this:

[ ] implement 16~~ functions
[ ] ...
[ ] compile yagna with your new driver
[ ] test it and verify it's working using instructions from 2)
[ ] send us a pull request


2) I am missing a section at the end about "testing / verify"

We have to give clear expectations on what do we consider "done". When a developer finishes writing a driver - how they can know that it works, and how we will be testing it. E.g. we will:
* we will try all payments commands on their driver and it should work
* we will try running the quickstart tutorial but choosing their driver and running blender example on top of it and the transactions should be exececuted correctly after work is done
* etc


and everything should go smoothly

File: introduction/payment-driver-getting-started.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Getting started to write L2 payment driver for Yagna
```
Comment (maaktweluit) [inline introduction/payment-driver-getting-started.md:1]: It does not have to be a L2 driver, they can also implement a paypal driver if they want ;)

File: introduction/payment-driver-getting-started.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+- communicate with daemon services over Golem Service Bus (GSB),
+- provide operation like deposits, exits, transfer, account unlocking,
+- delegate transaction's signing to the Identity service,
+- notify Payment service about committed transactions,
+- retry operations that failed to due non-critical causes.
+
+Here is a checklist that driver implementors would want to follow:
+
+- [] Fork the Yagna repository and clone it locally to start working on the code,
+- [] Implement all methods from `PaymentDriver` and `PaymentDriverCron` traits, some of the them are describe in the section below,
```
Comment (maaktweluit) [inline introduction/payment-driver-getting-started.md:20]: Cron is optional, Db is also available to be used

File: introduction/payment-driver-getting-started.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+## Testing a new driver
+
+When you consider your work is done, take some time to actually see it in action. Be sure we will scrupulously test the driver behavior in the following scenarios, including the error messages written in logs and returned from CLI. Don't hesitate to ask questions, we are here to help.  
+
+We will run the following tests (happy paths and providing invalid parameters to make it fail).
+
+- command line interaction with the driver: `init`, `fund`, `enter`, `transfer`, `status`,  `exit` ...
+- [quickstart tutorial](https://golem-network.gitbook.io/golem-sdk-develop/requestor-tutorials/flash-tutorial-of-requestor-development) with the blender example task,
+- review made transactions in the corresponding block explorer.
```
Comment (maaktweluit) [inline introduction/payment-driver-getting-started.md:81]: i think we should also add invoice_flow and debit_note flow tests in this list

**Review by maaktweluit — CHANGES_REQUESTED** [review]
Nice doc!

The most important comment is about the DB / Cron, will approve on reply or fix :)

File: introduction/payment-driver-getting-started.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+- communicate with daemon services over Golem Service Bus (GSB),
+- provide operation like deposits, exits, transfer, account unlocking,
+- delegate transaction's signing to the Identity service,
+- notify Payment service about committed transactions,
+- retry operations that failed to due non-critical causes.
+
+Here is a checklist that driver implementors would want to follow:
+
+- [] Fork the Yagna repository and clone it locally to start working on the code,
```
Comment (Wiezzel) [inline introduction/payment-driver-getting-started.md:19]: I'd add something like "create a new crate for your driver in `core/payment-drivers` directory.

File: introduction/payment-driver-getting-started.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+- [] Implement all methods from `PaymentDriver` trait, some of the them are describe in the section below,
+- [] Your implementation can also benefits from database handling code and backgroud tasks in `cron` module, but it is not mandatory to use them,
+- [] Make a new driver builds and starts with the Yagna daemon. See instructions bellow,
```
Comment (Wiezzel) [inline introduction/payment-driver-getting-started.md:22]:
```suggestion
- [] Make a new driver build and start with the Yagna daemon. See instructions bellow,
```

File: introduction/payment-driver-getting-started.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+- `verify_payment` - Verifies the payment transaction specified by transaction's confirmation (transaction's identifier).
+- `validate_allocation` - Validates that allocated funds are still sufficient to cover the costs of the task (including the transaction fees, e.g. Ethereum's Gas). Allocation is created when the requestor publishes the task on the market.
+- `account_event` - Called by the Identity service to notify the driver that specified account is _locked_ / _unlocked_. Identity service holds accounts private keys and signs transactions.
+- `recv_init_required` - Tells whether account initialization is needed for receiving payments.
+- `get_transaction_balance` - Deprecated. Drivers should return very big number (e.g. `1_000_000_000_000_000_000u64` or the whole token supply).
+
+
+The following driver's operations are processed periodically by the cron jobs. It is not mandatory to implemeny `PaymentDriverCron` trait as long as the driver can operate without the cron.
```
Comment (Wiezzel) [inline introduction/payment-driver-getting-started.md:59]: Please run the whole text through a spellchecker.
```suggestion
The following driver's operations are processed periodically by the cron jobs. It is not mandatory to implement `PaymentDriverCron` trait as long as the driver can operate without the cron.
```

File: introduction/payment-driver-getting-started.md (hunk 4 above)
Comment (Wiezzel) [reply introduction/payment-driver-getting-started.md:19]: Bump.

Comment (pnowosie) [reply introduction/payment-driver-getting-started.md:19]: :) just in the separate (atomic) commit

**Review by Wiezzel — APPROVED** [review]

**Comment by pnowosie** [comment]
@rad9k - I'm inexperienced in the git-book format. I hope the above is enough - but if you have some suggestion pls have your say.
Can we merge it right-away :question:

**Comment by rad9k** [comment]
I need to have a little thinking session as this (or at last part of the pr) seems to fit the into https://golem-network.gitbook.io/golem-infrastructure-documentation-develop/

I would like to progress with the thinking session tomorrow EOD

**Comment by pnowosie** [comment]
Ping @rad9k , @83tb  :pray:

**Review by rad9k — APPROVED** [review]
1) from my (limited) point of view this seems understable and detailed enough to get dev going into implementation of the driver
2) format is +- OK for the gitbook. for the final check need to merge and automatically process through gitbook and then do the corrections on the fly if needed
3) at this moment we are not referencing the tutorial from the gitbook table of content so we can have marge anytime, and I will do the corrections afterwards

**Review by 83tb — APPROVED** [review]
good to go

**Pull request merged**